CMakeLists.txt

```diff
@@ -258,6 +258,10 @@ if(ENABLE_DEBUG)
   set(DEBUGBUILD 1)
 endif()
 
+# Automatically enabled for shared libraries,
+# required for static library builds with ASAN.
+set(CMAKE_POSITION_INDEPENDENT_CODE ON)
```

Comment:
I'm not 100% sure but I would try setting this via a command line argument and see if that works before hardcoding it in the CMake file. `-DCMAKE_POSITION_INDEPENT_CODE=ON`

Reply:
These variables are not cache variables such as `CMAKE_BUILD_TYPE` and if I am not mistaken they are supposed to be set by the developer as opposed to the user.

Setting this option has two effects, enabling -fPIC for all libraries, and enabling PIE for executables (often seen as a hardening feature). I think there is value in enabling it by default (it changes the behavior for static libs and executables), but clearly some more work is needed as the macOS build somehow fails.

Docs for this option: https://cmake.org/cmake/help/latest/variable/CMAKE_POSITION_INDEPENDENT_CODE.html

Reply:
The docs for this stuff really need to be updated. I just tried it out on my own library, and specifying `-DCMAKE_POSITION_INDEPENDENT_CODE=ON` via a command line argument (using static libraries) works and keeps its value over multiple CMake invocations so it effectively behaves like an option.

Reply:
If you specify `-Dx=y` through the `cmake` invocation, it will be stored in the cache (as if [`set(x y CACHE STRING "..."`](https://cmake.org/cmake/help/latest/command/set.html#set-cache-entry) was called. The difference between a plain variable and an "option" is that "option" creates a cache entry by default such that you can select it through `ccmake` or `cmake-gui`.

I think I agree with your viewpoint. It is somewhat annoying if projects don't follow CMake standards and add `-O2 -DNDEBUG` even when using `-DCMAKE_BUILD_TYPE=None`. For top-level projects, setting `CMAKE_POSITION_INDEPENDENT_CODE` might seem reasonable, but if it is included via `add_subdirectory` as used in [bnl](https://github.com/DaanDeMeyer/bnl/blob/master/cmake/bnl.cmake#L290), it will probably indeed override stuff from the parent project. To handle such embedded cases better, I'l look into target-specific options.

Reply:
Maybe put this behind the `ENABLE_ASAN` option as well? I'm not sure if you need `-fPIC` when not compiling with sanitizers.

Reply:
Have a look at the new commit message, I believe that `-fPIC` is necessary whenever you try to combine a static library (ngtcp2) into another shared library (which is always built with `-fPIC`).